clippy_lints/src/derive.rs

```diff
@@ -432,6 +432,13 @@ impl<'tcx> Visitor<'tcx> for UnsafeVisitor<'_, 'tcx> {
     fn visit_expr(&mut self, expr: &'tcx Expr<'_>) -> Self::Result {
         if let ExprKind::Block(block, _) = expr.kind
             && block.rules == BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided)
+            && block
+                .span
+                .source_callee()
+                .and_then(|expr| expr.macro_def_id)
+                .is_none_or(|did| {
+                    (self.cx.tcx.crate_name(did.krate), self.cx.tcx.item_name(did)) != (sym::core, sym::pin)
+                })
```

Comment:
This seems oddly specific to `pin!`; what about something more general such as:

```suggestion
                .from_expansion()
```

  - or, if using a `LatePass`, https://doc.rust-lang.org/1.87.0/nightly-rustc/rustc_span/struct.Span.html#method.in_external_macro seems idoneous.